pkg/spdk/engine.go

```diff
@@ -1229,6 +1304,388 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+			}
+
+			e.lastExpansionError = errors.Wrap(err, "engine under non-recoverable operation").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		e.finishExpansion(expanded, size)
+	}()
+
+	// fetch all replica clients
+	replicaClients, err := e.getReplicaClients()
+	if err != nil {
+		return err
+	}
+	defer e.closeRplicaClients(replicaClients)
+	if len(replicaClients) == 0 || e.isAnyReplicaUnderRebuilding(replicaClients) {
+		return errors.New("expansion failed: no healthy replica or at least one replica is under rebuilding")
+	}
+
+	// prepareRaidForExpansion checks if RAID exists, suspends frontend if needed, and deletes the RAID bdev.
+	isSusupend, bdevRaidUUID, err := e.prepareRaidForExpansion(spdkClient)
+	if err != nil {
+		return errors.Wrap(err, "prepare raid for expansion failed")
+	}
+	if isSusupend {
+		defer func() {
+			if frontendErr := e.initiator.Resume(); frontendErr != nil {
+				frontendErr = errors.Wrapf(frontendErr, "failed to resume NVMe initiator during engine %s", e.Name)
+				engineErr = frontendErr
+			}
+		}()
+	}
+
+	// expand replicas
+	expansionSucceeded := e.expandReplicas(replicaClients, spdkClient, size)
+
+	// recreate RAID and reconnect frontend
+	if err := e.reconnectFrontend(spdkClient, bdevRaidUUID); err != nil {
+		e.log.WithFields(logrus.Fields{
+			"engineName": e.Name,
+			"volumeName": e.VolumeName,
+			"error":      err,
+		}).Error("reconnectFrontend failed")
+		return err
+	}
+
+	e.log.WithFields(logrus.Fields{
+		"engineName": e.Name,
+		"volumeName": e.VolumeName,
+		"frontend":   e.Frontend,
+		"specSize":   e.SpecSize,
+		"newSize":    size,
+	}).Info("Expanding engine complete")
+
+	expanded = expansionSucceeded // which could be true even in partial success
+	return nil
+}
+
+func (e *Engine) startExpansion(size uint64) (err error) {
+	if e.isExpanding {
+		return fmt.Errorf("expansion is in progress")
+	}
+
+	if e.IsRestoring {
+		return fmt.Errorf("restoring is in progress")
+	}
+
+	if types.IsUblkFrontend(e.Frontend) {
+		return errors.New("ublk frontend not support expansion")
+	}
+
+	defer func() {
+		if e.isExpanding {
+			e.lastExpansionFailedAt = ""
+			e.lastExpansionError = ""
+		} else if err != nil {
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+			e.lastExpansionError = errors.Wrap(err, "engine failed to start expansion").Error()
+		}
+	}()
+
+	if e.SpecSize > size {
+		return fmt.Errorf("cannot expand engine to a smaller size %v, current size %v", size, e.SpecSize)
+	} else if e.SpecSize == size {
+		e.log.Infof("Replica had been expanded to size %v", size)
+		return nil
+	}
+
+	roundedNewSize := util.RoundUp(size, helpertypes.MiB)
+	if roundedNewSize != size {
+		return fmt.Errorf("rounded up spec size from %v to %v since the spec size should be multiple of MiB", size, roundedNewSize)
+	}
+
+	e.isExpanding = true
```

Comment:
The size check should be executed before the `defer` error handling, since:
1. A wrong expansion request should not error the engine
2. Retry for an invalid size expansion does not make sense.